<a href="https://colab.research.google.com/github/murpunk/Programming_2023/blob/main/15_05_23_Metrics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#I. Прочитайте и повторите примеры из книги *Андреас Мюллер, Сара Гвидо. Введение в машинное обучение с помощью Python. Руководство для специалистов по работе с данными.* со страниц 295-320.


Пример: попытка построить модель, предсказывающую, кликнет пользователь на ссылку или нет. Нам известно, что он кликает только в 1% случаев из 100. Выборка получается несбалансированная по объёму примеров.

In [10]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
import numpy as np

In [2]:
digits = load_digits()

In [ ]:
digits

In [4]:
y = digits.target == 9

In [9]:
X_train, X_test, y_train, y_test = train_test_split(digits.data, y, random_state = 0)

In [12]:
from sklearn.dummy import DummyClassifier
dummy_majority = DummyClassifier(strategy = "most_frequent").fit(X_train, y_train)
pred_most_frequent = dummy_majority.predict(X_test)
print(f"Уникальные спрогнозированные метки: {np.unique(pred_most_frequent)}")
print(f"Правильность на тестовом наборе: {round(dummy_majority.score(X_test, y_test), 2)}")

Уникальные спрогнозированные метки: [False]
Правильность на тестовом наборе: 0.9


In [13]:
from sklearn.tree import DecisionTreeClassifier
tree = DecisionTreeClassifier(max_depth=2).fit(X_train, y_train)
pred_tree = tree.predict(X_test)
print(f"Правильность на тестовом наборе: {round(tree.score(X_test, y_test), 2)}")

Правильность на тестовом наборе: 0.92


In [17]:
from sklearn.linear_model import LogisticRegression
dummy = DummyClassifier().fit(X_train, y_train)
pred_dummy = dummy.predict(X_test)
print(f"Правильность dummy: {round(dummy.score(X_test, y_test), 2)}")
logreg = LogisticRegression(C=0.1).fit(X_train, y_train)
pred_logreg = logreg.predict(X_test)
print(f"Правильность logreg: {round(logreg.score(X_test, y_test), 2)}")

Правильность dummy: 0.9
Правильность logreg: 0.98


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [15]:
from sklearn.metrics import confusion_matrix
confusion = confusion_matrix(y_test, pred_logred)
print(f"Confusion matrix: \n{confusion}")

Confusion matrix: 
[[402   1]
 [  6  41]]


402 -- правильно отнесённые к отрицательным (TN)
6 -- положительные, которые были неверно отнесены к отрицательным (FN)
41 -- правильно отнесённые к положительным (TP)
1 -- отрицательные, неверно отнесённые  положительным (FP)

In [18]:
print("Наиболее часто встречающийся класс:")
print(confusion_matrix(y_test, pred_most_frequent))
print("\nDummy-модель:")
print(confusion_matrix(y_test, pred_dummy))
print("\nДерево решений:")
print(confusion_matrix(y_test, pred_tree))
print("\nЛогистическая регрессия:")
print(confusion_matrix(y_test, pred_logreg))

Наиболее часто встречающийся класс:
[[403   0]
 [ 47   0]]

Dummy-модель:
[[403   0]
 [ 47   0]]

Дерево решений:
[[390  13]
 [ 24  23]]

Логистическая регрессия:
[[402   1]
 [  6  41]]


## Метрики

**Правильность**: $\frac{TP + TN}{TP + TN + FP + FN}$


То есть количество верно классифицированных примеров на общее количество примеров.

**Точность** (precision): $\frac{TP}{TP + FP}$


Сколько из предсказанных положительных примеров оказались действительно положительными. Точность важна, когда необходимо снизить количество ложноположительных примеров.


Эта метрика ещё называется как **прогностическая ценность положительного результата**: PPV.

**Полнота** (recall): $\frac{TP}{TP + FN}$

Полнота показывает, сколько от числа фактических положительных примеров было предсказано как полодительный класс.

Метрика полноты важна, если нужно сократить количество ложноотрицательных примеров.